src/toil/batchSystems/mesos/executor.py

```diff
@@ -98,7 +100,10 @@ def launchTask(self, driver, task):
         def runTask():
             log.debug("Running task %s", task.task_id.value)
             sendUpdate(mesos_pb2.TASK_RUNNING)
-            popen = runJob(pickle.loads(task.data))
+            # This is where task.data is first invoked. Using this position to setup cleanupInfo
+            taskData = pickle.loads(task.data)
+            self.workerCleanupInfo = taskData.workerCleanupInfo
```

Comment:
Make the assignment conditional (so it only happens once) and assert that subsequent tasks don't change it.